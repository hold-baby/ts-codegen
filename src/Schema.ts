import { getUseExtends, isArray, toCapitalCase } from "src/utils";
import { forEach, isEmpty, map, reduce } from "lodash";
import { IReference, ISchema } from "src/__types__/OpenAPI";
import { CustomType, Type } from "src/Type";
import { CustomSchema } from "src/__types__/types";
import { createRegister } from "src/Register";

export class Schema {
  private type: Type;

  constructor(register: ReturnType<typeof createRegister>) {
    this.type = new Type(register);
  }

  public convert(schema: CustomSchema, id?: string): CustomType {
    const name = id ? toCapitalCase(id) : id;
    const oneOf = (schema as ISchema).oneOf || (schema as ISchema).anyOf;
    if (oneOf) {
      return this.type.oneOf(map(oneOf, (v) => this.convert(v)));
    }

    const allOf = (schema as ISchema).allOf;
    if (allOf) {
      const { props, refs, useExtends } = this.handleAllOf(allOf, name);
      return this.type.object(props, refs, useExtends);
    }

    if (schema.items) {
      return this.type.array(this.handleItems(schema.items, name));
    }

    if (schema.$ref) {
      return this.type.ref((schema as IReference).$ref);
    }

    if (schema.enum) {
      return this.type.enum(schema.enum, name);
    }

    if (schema.type === "object" || schema.properties) {
      return schema.properties ? this.type.object(this.handleObject(schema, name)) : this.type.object("object");
    }

    if (schema.type === "string") {
      return this.type.string();
    }

    if (schema.type === "boolean") {
      return this.type.boolean();
    }

    if (schema.type === "integer" || schema.type === "number") {
      return this.type.number();
    }

    if (schema.type === "file") {
      return this.type.file();
    }

    return this.type.null();
  }

  private handleAllOf(schemas: Array<CustomSchema>, name?: string) {
    const refs: any[] = [];
    let props: any = {};
    let useExtends = getUseExtends(schemas);

    forEach(schemas, (schema) => {
      if (schema.$ref) {
        refs.push(this.convert(schema, name));
      } else if (!isEmpty(schema)) {
        props = this.convert(schema, name);
      }
    });

    return {
      refs,
      props,
      useExtends,
    };
  }

  private handleItems(items: CustomSchema | IReference | CustomSchema[], name?: string): CustomType | CustomType[] {
    if (isArray(items)) {
      return map(items, (v) => this.handleItems(v, name)) as CustomType[];
    }
    return this.convert(items, name);
  }

  private handleObject(schema: CustomSchema, name?: string): { [key: string]: CustomType } {
    return reduce(
      schema.properties,
      (res, v, k) => {
        const isRequired = (v as CustomSchema)?.required || schema.required?.includes(k);
        return {
          ...res,
          [`${k}${isRequired ? "" : "?"}`]: this.convert(v, `${name}${toCapitalCase(k)}`),
        };
      },
      {},
    );
  }
}
