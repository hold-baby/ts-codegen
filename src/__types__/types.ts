import { CustomType } from "src/Type";
import { Operation, Parameter, Path, Schema, Spec } from "swagger-schema-official";
import { IComponents, IOpenAPI, IOperation, IPathItem, ISchema, TParameter } from "src/__types__/OpenAPI";

export type RequestType = { [key: string]: CustomType };

export interface IClientConfig {
  url: string;
  method: string;
  TResp: CustomType | undefined;
  TReq: RequestType | { [key: string]: RequestType } | undefined;
  operationId?: string;
  pathParams: string[];
  queryParams: string[];
  bodyParams?: string[]; // for backward capability, only used in swagger version 2.0
  contentType?: string;
  deprecated?: boolean;
}

export type CustomSchema = Schema | ISchema;
export type CustomReference = {
  $ref: string;
};
export type CustomParameters = Operation["parameters"] | IOperation["parameters"];
export type CustomResponses = Spec["responses"] | IComponents["responses"];
export type CustomParameter = Parameter | TParameter;
export type CustomPath = Path | IPathItem;
export type CustomPaths = Spec["paths"] | IOpenAPI["paths"];
export type CustomOperation = Operation | IOperation;
