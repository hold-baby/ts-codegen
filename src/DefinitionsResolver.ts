import { Schema } from "swagger-schema-official";
import { addPrefixForInterface, arrayToObject, isNumber, toCapitalCase, toTypes } from "./utils";
import { compact, Dictionary, forEach, includes, replace, some } from "lodash";
import { SchemaResolver2 } from "src/swagger/v3/SchemaResolver2";

// TODO: 1. Handle required params.
// TODO: handle `in: fromData`
// TODO: handle `in schema`
// TODO: 确认不同 endpoint 是否都会生成 swagger

const ENUM_SUFFIX = `#EnumTypeSuffix`;

export function generateEnums(data: Dictionary<any>, key: string) {
  if (!data) {
    return "";
  }

  const enums = data[key];
  const hasNumber = some(enums, (v) => isNumber(v));
  const enumName = replace(key, ENUM_SUFFIX, "");
  return hasNumber
    ? `export type ${enumName} = ${enums.map((item: string | number) => JSON.stringify(item)).join("|")}`
    : `export enum ${enumName} ${JSON.stringify(arrayToObject(enums)).replace(/:/gi, "=")}`;
}

export class DefinitionsResolver {
  public resolvedDefinitions: Dictionary<any> = {};

  static of(definitions?: { [definitionsName: string]: Schema }) {
    return new DefinitionsResolver(definitions);
  }

  constructor(private definitions?: { [definitionsName: string]: Schema }) {}

  scan = () => {
    const r = SchemaResolver2.of((k, v) => {
      this.resolvedDefinitions[k] = v;
    });

    forEach(this.definitions, (v, k) =>
      r.resolve({
        ...v,
        _name: k,
        _propKey: k,
      }),
    );

    return this;
  };

  toDeclarations = (): string[] => {
    const resolvedDefinitions = this.resolvedDefinitions;
    const arr = Object.keys(resolvedDefinitions)
      .sort()
      .map((key) => {
        if (includes(key, ENUM_SUFFIX)) {
          return generateEnums(resolvedDefinitions, key);
        }

        if (resolvedDefinitions[key] === "object") {
          return `export interface ${addPrefixForInterface(toCapitalCase(key))} {[key:string]:any}`;
        }
        const val = toTypes(resolvedDefinitions[key]);
        if (val) {
          return `export interface ${addPrefixForInterface(toCapitalCase(key))} ${val}`;
        }
      });

    return compact(arr);
  };
}
