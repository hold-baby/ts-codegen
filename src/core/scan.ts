import { IReference } from "src/v3/OpenAPI";
import { CustomSchema, Ref, Register } from "src/core/Type";
import { keys } from "lodash";
import { getUseExtends, Schema } from "src/core/Schema";
import { prettifyCode, toCapitalCase } from "src/core/utils";

const getDeclarationType = (schema: CustomSchema) => {
  if (schema.type === "object" || schema.properties || (schema.allOf && getUseExtends(schema.allOf))) {
    return "interface";
  }
  return "type";
};

const addPrefix = (name: string) => `${Register.prefixes[name] === "interface" ? "I" : "T"}${name}`;

export const scan = (schemas: { [k: string]: CustomSchema | IReference }) => {
  const schemaHandler = new Schema();

  keys(schemas).forEach((k) => {
    const name = toCapitalCase(k);
    const type = schemaHandler.convert(schemas[k], name);
    Register.setType(name, type);
    Register.setPrefix(name, getDeclarationType(schemas[k]));
  });

  for (let name in Register.refs) {
    (Register.refs[name] as Ref).rename(addPrefix(name));
  }

  let output = "";
  keys(Register.decls)
    .sort()
    .forEach((k) => {
      output =
        output +
        `export ${Register.prefixes[k]} ${addPrefix(k)} ${
          Register.prefixes[k] === "interface" ? "" : "="
        } ${Register.decls[k].toType()}${Register.prefixes[k] === "type" ? ";" : ""}\n\n`;
    });

  return prettifyCode(output);
};
