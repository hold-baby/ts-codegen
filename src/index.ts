export * from "src/v2/ClientConfigs";
export * from "src/v3/ClientConfigsV3";
export * from "src/core/utils";
