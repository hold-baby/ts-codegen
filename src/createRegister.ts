// 利用闭包持有状态（私有变量）
import { CustomType, Ref } from "src/Type";
import { Parameter } from "swagger-schema-official";
import { IReference, IRequestBody, IResponse } from "src/__types__/OpenAPI";
import { get, set } from "lodash";

export enum DeclKinds {
  interface = "interface",
  type = "type",
  enum = "enum",
}

const withPrefix = (name: string, kind: string) => {
  switch (kind) {
    case "interface":
      return `I${name}`;
    case "type":
      return `T${name}`;
    default:
      return name;
  }
};

export interface IStore {
  decls: {
    [id: string]: {
      type: CustomType;
      kind: string;
      name: string;
    };
  };
  refs: { [id: string]: Ref };
  parameters: {
    [id: string]: Parameter;
  };
  responses: {
    [id: string]: Response | IResponse | IReference;
  };
  requestBodies: {
    [id: string]: IReference | IRequestBody;
  };
}

export const createRegister = () => {
  const store: IStore = {
    decls: {},
    refs: {},
    parameters: {},
    responses: {},
    requestBodies: {},
  };

  return {
    getDecls() {
      return store.decls;
    },

    setDecl: (id: string, type: CustomType, kind: string) => {
      store.decls[id] = {
        type,
        kind,
        name: withPrefix(id, kind),
      };
    },

    setRef: (id: string) => {
      if (store.refs[id]) {
        return store.refs[id];
      }

      const type = new Ref(id);
      store.refs[id] = type;

      return type;
    },

    getData(paths: string[]) {
      return get(store, paths);
    },

    setData(paths: string[], data: any) {
      return set(store, paths, data);
    },

    renameAllRefs: (cb: (newName: string) => string) => {
      for (let name in store.refs) {
        store.refs[name].rename(cb(name));
      }
    },
  };
};
