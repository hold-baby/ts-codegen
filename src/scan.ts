import { IOpenAPI, IServer } from "src/__types__/OpenAPI";
import { CustomType, Enum } from "src/Type";
import { compact, get, isEmpty, keys, mapValues, sortBy } from "lodash";
import { Schema } from "src/Schema";
import { getUseExtends, prettifyCode, setDeprecated, toCapitalCase, toTypes } from "src/utils";
import { Spec } from "swagger-schema-official";
import { CustomReference, CustomSchema, IClientConfig } from "src/__types__/types";
import { getClientConfigsV2, getClientConfigsV3 } from "src/index";
import { createRegister } from "src/Register";
import { parse } from "url";

const isOpenApi = (v: any): v is IOpenAPI => v.openapi;

enum DataType {
  openapi,
  swagger,
}

const getInputs = (data: Spec | IOpenAPI) => {
  if (isOpenApi(data)) {
    return {
      type: DataType.openapi,
      basePath: getBasePathFromServers(data?.servers),
      paths: data.paths,
      schemas: data.components?.schemas as { [key: string]: CustomSchema | CustomReference },
      parameters: data.components?.parameters,
      responses: data.components?.responses,
      requestBodies: data.components?.requestBodies,
    };
  }
  return {
    type: DataType.swagger,
    basePath: data.basePath || "",
    paths: data.paths,
    schemas: data.definitions as { [key: string]: CustomSchema },
    parameters: data.parameters,
    responses: data.responses,
    requestBodies: null,
  };
};

export const scan = (data: Spec | IOpenAPI) => {
  const register = createRegister();
  const schemaHandler = new Schema(register);
  const { type, basePath, paths, schemas, parameters, responses, requestBodies } = getInputs(data);

  function toReusableTypes(s: { [k: string]: CustomSchema | CustomReference }) {
    return keys(s).map((k) => {
      const name = toCapitalCase(k);
      const t = schemaHandler.convert(s[k], name);
      register.setType(name, t);
      register.setPrefix(name, getDeclarationType(s[k]));
      return t;
    });
  }

  toReusableTypes(schemas);

  register.setData(["parameters"], parameters);
  register.setData(["responses"], responses);
  register.setData(["requestBodies"], requestBodies);

  let clientConfigs: IClientConfig[] =
    type === DataType.swagger
      ? getClientConfigsV2(paths, basePath, register)
      : getClientConfigsV3(paths, basePath, register);

  const decls = register.getDecls();
  const prefixes = register.getPrefixes();
  register.renameAllRefs((name) => addPrefix(name, prefixes));

  return prettifyCode(`${toRequest(clientConfigs)} \n\n ${getOutput(decls, prefixes)}`);
};

function toRequest(clientConfigs: IClientConfig[]): string {
  // for (let name in Register.refs) {
  //   if (!(Register.refs[name] as Ref).alias) {
  //     (Register.refs[name] as Ref).rename(addPrefix(name));
  //   }
  // }
  const clientConfig = sortBy(clientConfigs, (o) => o.operationId);

  // TODO: refactor code later
  function mapper(obj: KType | { [key: string]: KType }): any {
    return toTypes(
      mapValues(obj, (v: any) => {
        if (!v.toType) {
          return mapper(v);
        }
        return v.toType(false);
      }),
    );
  }

  return clientConfig
    .map((v: IClientConfig) => {
      const TReq = !isEmpty(v.TReq) ? mapper(v.TReq as any) : "";
      const requestParamList = compact([...v.pathParams, ...v.queryParams, v.contentType ? "requestBody" : ""]);
      const requestInputs = isEmpty(requestParamList) ? "" : toRequestParams(requestParamList);

      const getParams = () => {
        const params = toRequestParams(get(v, "queryParams"));
        return params ? `params: ${params},` : "";
      };

      const getHeaders = () => (v.contentType ? `headers: { "Content-Type": '${v.contentType}' },` : "");

      const types = compact([TReq, v.TResp?.toType(false)]).join(",");
      return `
${v.deprecated ? setDeprecated(v.operationId) : ""}
export const ${v.operationId} = createRequestAction${types ? "<" + types + ">" : ""}("${
        v.operationId
      }", (${requestInputs}) => ({ url: \`${v.url}\`,method: "${v.method}",${
        v.contentType ? `data: requestBody,` : ""
      }${getParams()}${getHeaders()} })
);
`;
    })
    .join("\n\n");
}

const toRequestParams = (data: any[] = []) =>
  !isEmpty(data)
    ? `{
${data.join(",\n")}
}`
    : undefined;

const getDeclarationType = (schema: CustomSchema) => {
  if (schema.type === "object" || schema.properties || (schema.allOf && getUseExtends(schema.allOf))) {
    return "interface";
  }
  return "type";
};

const addPrefix = (name: string, prefixes: { [id: string]: string }) =>
  `${prefixes[name] === "interface" ? "I" : "T"}${name}`;

const getOutput = (decls: { [id: string]: CustomType }, prefixes: { [id: string]: string }): string => {
  let output = "";
  keys(decls)
    .sort()
    .forEach((k) => {
      const t = decls[k];
      if (t instanceof Enum) {
        output = output + `export ${t.toType()}\n\n`;
        return;
      }

      output =
        output +
        `export ${prefixes[k]} ${addPrefix(k, prefixes)} ${prefixes[k] === "interface" ? "" : "="} ${t.toType()}${
          prefixes[k] === "type" ? ";" : ""
        }\n\n`;
    });
  return output;
};

type KType = { [key: string]: CustomType };

const getBasePathFromServers = (servers?: IServer[]): string => {
  if (isEmpty(servers)) {
    return "";
  }
  const server = servers![0];
  if (server?.variables) {
    const basePath = get(server, "variables.basePath.default");
    return basePath ? `/${basePath}` : "";
  }
  return parse(server?.url)?.pathname || "";
};
