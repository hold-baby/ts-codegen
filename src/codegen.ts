import { getFilename, prettifyCode, testJSON } from "src/utils";
import axios from "axios";
import { isEmpty, map } from "lodash";
import * as fs from "fs";
import * as path from "path";
import { IOpenAPI } from "src/__types__/OpenAPI";
import { Spec } from "swagger-schema-official";
import { getInputs, scan } from "src/scan";
import { ERROR_MESSAGES } from "src/constants";

export const getCodegenConfig = () => {
  const codegenConfigPath = path.resolve("ts-codegen.config.json");
  return fs.existsSync(codegenConfigPath)
    ? require(codegenConfigPath)
    : {
        output: ".output",
        actionCreatorImport: "",
        clients: [],
        options: {
          typeWithPrefix: false,
          backwardCompatible: false,
        },
      };
};

export const codegen = () => {
  const { output, actionCreatorImport, timeout, data, clients, options } = getCodegenConfig();

  const writeSpecToFile = (spec: IOpenAPI | Spec) => {
    if (!spec) {
      return;
    }
    const fileStr = `${actionCreatorImport} ${scan(spec, options)}`;
    const { basePath } = getInputs(spec);
    write(output, `./${getFilename(basePath)}`, fileStr);
  };

  if (!isEmpty(data)) {
    data.map((file: string) => {
      const specStr = fs.readFileSync(file, "utf8");
      const spec = testJSON(specStr, ERROR_MESSAGES.INVALID_JSON_FILE_ERROR);

      writeSpecToFile(spec);
    });
  }

  if (!isEmpty(clients)) {
    fetchSwaggerJSON(clients, timeout).then((results: any[]) => {
      results.forEach((spec: IOpenAPI | Spec) => {
        writeSpecToFile(spec);
      });
    });
  }
};

const write = (output: string, filename: string, str: string) => {
  if (!fs.existsSync(output)) {
    fs.mkdirSync(output);
  }

  fs.writeFileSync(path.resolve(output, `./${filename}.ts`), prettifyCode(str), "utf-8");
};

const fetchSwaggerJSON = (clients: string[] = [], timeout: number = 10 * 1000) => {
  const instance = axios.create({
    timeout,
  });

  return Promise.all(
    map(clients, (client) => {
      instance
        .get(client)
        .then((response) => response.data)
        .catch((error) => {
          console.error(`${error.code}: ${ERROR_MESSAGES.FETCH_CLIENT_FAILED_ERROR}`);
        });
    }),
  );
};
