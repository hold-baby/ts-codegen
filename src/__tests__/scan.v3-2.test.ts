import swaggerV3PetStore from "examples/petstore.v3.json";
import { scan } from "src/scan";
import { IOpenAPI } from "src/__types__/OpenAPI";

describe("swagger v3", () => {
  it("should handle special schemas correctly", () => {
    expect(scan(swaggerV3PetStore as IOpenAPI)).toMatchSnapshot();
  });
});
