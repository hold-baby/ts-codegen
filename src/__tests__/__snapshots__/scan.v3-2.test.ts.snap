// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`swagger v3 should handle special schemas correctly 1`] = `
"export const addPet = createRequestAction<
  {
    requestBody: INewPet;
  },
  TPet
>(\\"addPet\\", ({ requestBody }) => ({
  url: \`/v2/pets\`,
  method: \\"post\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const deletePet = createRequestAction<{
  id: number;
}>(\\"deletePet\\", ({ id }) => ({ url: \`/v2/pets/\${id}\`, method: \\"delete\\" }));

export const findPetById = createRequestAction<
  {
    id: number;
  },
  TPet
>(\\"findPetById\\", ({ id }) => ({ url: \`/v2/pets/\${id}\`, method: \\"get\\" }));

export const findPets = createRequestAction<
  {
    limit?: number;
    tags?: keyof typeof FindPetsTags[];
  },
  TPet[]
>(\\"findPets\\", ({ tags, limit }) => ({
  url: \`/v2/pets\`,
  method: \\"get\\",
  params: {
    tags,
    limit,
  },
}));

export const formMultipartWithFile = createRequestAction<{
  requestBody: { data?: TPet; file: string; slice?: string[]; string?: string };
}>(\\"formMultipartWithFile\\", ({ requestBody }) => ({
  url: \`/v2/demo/forms/multipart\`,
  method: \\"post\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"multipart/form-data\\" },
}));

export const formMultipartWithFiles = createRequestAction<{
  requestBody: { files: string[] };
}>(\\"formMultipartWithFiles\\", ({ requestBody }) => ({
  url: \`/v2/demo/forms/multipart-with-files\`,
  method: \\"post\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"multipart/form-data\\" },
}));

export const formUrlEncoded = createRequestAction<{
  requestBody: IRequestForm;
}>(\\"formUrlEncoded\\", ({ requestBody }) => ({
  url: \`/v2/demo/forms/url-encoded\`,
  method: \\"post\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"application/x-www-form-urlencoded\\" },
}));

export interface IError {
  code: number;
  message: string;
}

export enum FindPetsTags {
  \\"z\\" = \\"z\\",
  \\"b\\" = \\"b\\",
  \\"c\\" = \\"c\\",
  \\"a\\" = \\"a\\",
}

export interface INewPet {
  name: string;
  tag?: string;
}

export type TPet = INewPet & { id: number };

export interface IRequestForm {
  data: TPet;
  slice: string[];
  string: string;
}
"
`;
